Sources/NIO/Selector.swift

```diff
@@ -494,13 +494,23 @@ internal extension Selector where R == NIORegistration {
         }
 
         let futures: [EventLoopFuture<Void>] = self.registrations.map { (_, reg: NIORegistration) -> EventLoopFuture<Void> in
+            // The futures will only be notified (of success) once also the closeFuture of each Channel is notified.
+            // This only happens after all other actions on the Channel is complete and all events are propagated through the
+            // ChannelPipeline. We do this to minimize the risk to left over any tasks / promises that are tied to the
+            // EventLoop itself.
             switch reg {
             case .serverSocketChannel(let chan, _):
-                return chan.close()
+                return chan.close().then {
+                    chan.closeFuture
+                }
             case .socketChannel(let chan, _):
-                return chan.close()
+                return chan.close().then {
+                    chan.closeFuture
+                }
             case .datagramChannel(let chan, _):
-                return chan.close()
+                return chan.close().then {
+                    chan.closeFuture
+                }
```

Comment:
This has the effect of firing the future early if the close errors. We probably don't want to do that: presumably we should just `close()` and then return `chan.closeFuture`, rather than chaining through `then`?

Reply:
@Lukasa yeah you are right... let me do this.